db/sqlinterfaces.c

```diff
@@ -3140,22 +3194,49 @@ int sqlengine_prepare_engine(struct sqlthdstate *thd,
     return rc;
 }
 
-static void normalize_stmt_and_store(
-  struct sqlclntstate *clnt,
-  struct sql_state *rec
+static void free_normalized_sql(
+  struct sqlclntstate *clnt
 ){
   if (clnt->work.zNormSql) {
     free(clnt->work.zNormSql);
     clnt->work.zNormSql = 0;
   }
-  assert(rec && rec->stmt);
-  assert(rec && rec->sql);
+}
+
+static void free_original_normalized_sql(
+  struct sqlclntstate *clnt
+){
+  if (clnt->work.zOrigNormSql) {
+    free(clnt->work.zOrigNormSql);
+    clnt->work.zOrigNormSql = 0;
+  }
+}
+
+static void normalize_stmt_and_store(
+  struct sqlclntstate *clnt,
+  struct sql_state *rec
+){
   if (gbl_fingerprint_queries) {
-    const char *zNormSql = sqlite3_normalized_sql(rec->stmt);
-    if (zNormSql) {
-      clnt->work.zNormSql = strdup(zNormSql);
-    } else if (gbl_verbose_normalized_queries) {
-      logmsg(LOGMSG_USER, "FAILED sqlite3_normalized_sql({%s})\n", rec->sql);
+    if (rec != NULL) {
+      assert(rec->stmt);
+      assert(rec->sql);
+      const char *zNormSql = sqlite3_normalized_sql(rec->stmt);
+      if (zNormSql) {
+        assert(clnt->work.zNormSql==0);
+        clnt->work.zNormSql = strdup(zNormSql);
+      } else if (gbl_verbose_normalized_queries) {
+        logmsg(LOGMSG_USER, "FAILED sqlite3_normalized_sql({%s})\n", rec->sql);
+      }
+    } else {
+      assert(clnt->work.zSql);
```

Comment:
Add comments about the nature of the SQL statement in question here...  i.e. EXEC PROCEDURE (no prepare) versus other DML (SELECT, etc) ...